Adds new terminal commands and overhauls parts of NTOS

Ithalan opened the pull request:
This PR mainly adds a bunch of new terminal commands in an attempt to make usage of it a bit more interesting, and also let more people interact with this system without getting full access to all the toys.

In addition to this, there's also an overhaul to parts of NTOS, mainly relating to starting and stopping programs, saving files to disks and using network cards.

Changelog:

:cl: Ithalan
rscadd: Newly-added access_network_admin has replaced access_network, and access_network is now used to give a lesser degree of access to NTNet through terminal commands than admins have. Jobs that previously had access_network now have access_network_admin as well, and access_network has been added to engineers, roboticists and forensic investigators.

rscadd: Several new terminal commands have been added and some existing ones have been revised. Details on available commands can be accessed through the 'man' command as usual.

rscadd: Individual files or programs on a hard drive or portable drive can be flagged as invisible through terminal-usage, hiding them from view in most NTOS applications.

rscadd: Entire drives can be made read-only through terminal-usage. Files on read-only drives cannot be edited or deleted, and new files cannot be created.

tweak: Network cards can be given custom identifiers through terminal-usage, for easier recognition on NTNet.

tweak: Programs set to autorun can now have autorun toggled off by clicking the autorun button again.

tweak: When wireless access for specific NTNet functions is disabled, the connection is only rejected if the last device in a proxy chain is a wireless device.

balance: access restrictions are now bypassed for the ssh terminal command when used from an emagged computer. Antag hackers rejoice!

bugfix: The number of programs that a modular computer can run simultaneously is once again limited by the installed CPU.
/:cl:

A summary of the terminal commands now available below. Most are only restricted by the skill-requirement to use them, as they cannot affect any device other than the one they are executed on in most cases. access_network_admin works as well in all cases where access_network is required

-- Unskilled --
Using the terminal while unskilled still always yields critical failures, but some bugs and inconsistent behavior in their effects have been fixed, and new largely inconsequential types of failure have replaced the cases where before there was no feedback at all on the input.

-- Basic --

- clear - Clears the terminal history

- exit - Closes the terminal

- file - Basic file management, including on removable drives. Notable for being the only way manage hidden files, and toggle the hidden flag on files.

- man - Shows information about all available commands and their usage

- shutdown - Shuts down the device that the terminal is running on

-- Trained --

- ban - Manages the NTNet list of banned network IDs. People with **access_network** can only see the list with this command, while people with **access_network_admin** can also add and remove network IDs from the list

- hwinfo - Shows information about hardware installed in the device the terminal is running on

- lock - Allows toggling read-only mode on and off on hard drives and removable drives

- ping - Pings NTNet or another device to show if it is reachable. If the current device connects through one or more proxy devices, the entire proxy chain is also show.

- prog - Shows running programs on the current device, and allows programs to be stopped, started, shifted between foreground and background, and also set the autorun flag

-- Experienced --

- ifconfig - Shows information about the network interface on this device, and allows the identification string on it to be renamed. Also shows active remote terminal connections to or from this device.

- log - Sends the NTNet log to the entered email address. Requires **access_network**, but users without **access_network_admin** have their use of this command logged to the NTNet log as well

- proxy - Selects another network ID to use as a proxy for this one, with all outgoing net activity being shown as coming from the proxy device. If the proxy device is itself set up to use another device as a proxy, everything will be forwarded to that as well. Checks are in place to prevent circular proxy chains, and if one device in the chain becomes unreachable, all devices behind it lose net connectivity.

- relays - Shows information about the NTNet relay(s), including the are they are located in. Requires **network_access** to use.

- ssh - Opens a remote terminal on the NTNet-connected device belonging to the entered ID, allowing any terminal command except ssh itself to be executed on that device as if the user was using it directly. Requires **access_network_admin** to use, unless used from a computer that is emagged.

- status - Shows whether NTNet wifi and Intrusion Detection is enabled, and whether an intrusion alert has been trigger. Requires **access_network**

-- Master --

- locate - Shows the area in which an NTNet-connected device belonging to the entered ID is located. Requires **access_network**, but usage is logged to the NTNet log without **access_network_admin**

- purge - Resets the NTNet intrusion alarm and purges the log. Requires **acccess_network_admin** to use

- sysnotify - makes the device play a sound and show a notification that is visible to anyone nearby it.

SierraKomodo commented on code/datums/uplink/uplink_sources.dm:
```diff
@@ -27,11 +27,8 @@ GLOBAL_LIST_INIT(default_uplink_source_priority, list(
-	if(!P.hard_drive.try_store_file(program))
+	if(!P.hard_drive.save_file(program))
 		return SETUP_FAILED	//Not enough space or other issues.
-	P.hard_drive.store_file(program)
```
Was removing the call to `store_file()` here intentional?

Ithalan replied:
Yes. hard_drive.save_file() covers everything that the removes lines attempted.

SierraKomodo commented on code/game/jobs/access_datum.dm:
```diff
@@ -253,7 +253,13 @@
+	desc = "Network Operations"
+	region = ACCESS_REGION_RESEARCH
+
+/var/const/access_network_admin = "ACCESS_NETWORK_ADMIN" //43
```
The numbers are irrelevant and only existed for debugging reasons back when we switched from numeric access flags to strings.
```suggestion
/var/const/access_network_admin = "ACCESS_NETWORK_ADMIN"
```

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -42,9 +42,11 @@ var/global/datum/ntnet/ntnet_global = new()
 	add_log("NTNet logging system activated.")
 
-/datum/ntnet/proc/add_log_with_ids_check(var/log_string, var/obj/item/stock_parts/computer/network_card/source = null)
+/datum/ntnet/proc/add_log_with_ids_check(var/log_string, var/obj/item/stock_parts/computer/network_card/source = null, var/intrusion = 1)
```
- `var/` is redundant in parameters.
- Use `TRUE`/`FALSE` instead of `1`/`0` for boolean values.

```suggestion
/datum/ntnet/proc/add_log_with_ids_check(var/log_string, var/obj/item/stock_parts/computer/network_card/source = null, intrusion = TRUE)
```

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -42,9 +42,11 @@ var/global/datum/ntnet/ntnet_global = new()
-		add_log(log_string, source)
+		add_log("IDS WARNING: " + log_string, source)
+		if (intrusion)
+			intrusion_detection_alarm = 1
```
```suggestion
			intrusion_detection_alarm = TRUE
```

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -42,9 +42,11 @@ var/global/datum/ntnet/ntnet_global = new()
+/datum/ntnet/proc/add_log_with_ids_check(var/log_string, var/obj/item/stock_parts/computer/network_card/source = null, var/intrusion = 1)
 	if(intrusion_detection_enabled)
-		add_log(log_string, source)
+		add_log("IDS WARNING: " + log_string, source)
```
Shouldn't this log only appear if `intrusion` is also true?

Ithalan replied:
IDS-related log entries can be added without it necessarily being desired to trigger the intrusion alarm. The intrusion flag is just a handy shortcut for doing it while adding the log entry, since that covers the majority of the uses of the IDS functionality that I could find.

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -93,13 +95,12 @@ var/global/datum/ntnet/ntnet_global = new()
 
-// Checks whether NTNet operates. If parameter is passed checks whether specific function is enabled.
-/datum/ntnet/proc/check_function(var/specific_action = 0)
+// Checks whether NTNet operates.
```
Since we're editing this comment anyway, may as well make it match dmdoc format - https://github.com/Baystation12/Baystation12/wiki/Guide-to-SpacemanDMM-Docblocks

```suggestion
/// Checks whether NTNet operates.
```

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -108,17 +109,21 @@ var/global/datum/ntnet/ntnet_global = new()
-		return (operating && setting_systemcontrol)
 	return operating
 
+// Checks whether NTNet allows a specific action over wireless connections
```
```suggestion
/// Checks whether NTNet allows a specific action over wireless connections
```

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -108,17 +109,21 @@ var/global/datum/ntnet/ntnet_global = new()
 	return operating
 
+// Checks whether NTNet allows a specific action over wireless connections
+/datum/ntnet/proc/check_capability(var/specific_action)
```
```suggestion
/datum/ntnet/proc/check_capability(specific_action)
```

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -108,17 +109,21 @@ var/global/datum/ntnet/ntnet_global = new()
+			return setting_communication
+		if(NTNET_SYSTEMCONTROL)
+			return setting_systemcontrol
+	return 0
```
```suggestion
	return FALSE
```

SierraKomodo commented on code/modules/modular_computers/NTNet/emails/email_account.dm:
```diff
@@ -39,6 +41,9 @@
 
 /datum/computer_file/data/email_account/proc/send_mail(var/recipient_address, var/datum/computer_file/data/email_message/message, var/relayed = 0)
+	if(suspended)
+		return 0
```
```suggestion
		return FALSE
```

SierraKomodo commented on code/modules/modular_computers/NTNet/emails/email_account.dm:
```diff
@@ -49,10 +54,10 @@
 
 	if(!recipient.receive_mail(message, relayed))
-		return
+		return 0
```
```suggestion
		return FALSE
```

SierraKomodo commented on code/modules/modular_computers/NTNet/emails/email_account.dm:
```diff
@@ -49,10 +54,10 @@
 
 	outbox.Add(message)
-	ntnet_global.add_log_with_ids_check("EMAIL LOG: [login] -> [recipient.login] title: [message.title].")
+	ntnet_global.add_log_with_ids_check("EMAIL LOG: [login] -> [recipient.login] title: [message.title].", intrusion = 0)
```
```suggestion
	ntnet_global.add_log_with_ids_check("EMAIL LOG: [login] -> [recipient.login] title: [message.title].", intrusion = FALSE)
```

SierraKomodo commented on code/modules/modular_computers/file_system/computer_file.dm:
```diff
@@ -7,6 +7,7 @@ var/global/file_uid = 0
 	var/obj/item/stock_parts/computer/hard_drive/holder	// Holder that contains this file.
 	var/unsendable = 0										// Whether the file may be sent to someone via NTNet transfer or other means.
 	var/undeletable = 0										// Whether the file may be deleted. Setting to 1 prevents deletion/renaming/etc.
+	var/hidden = 0											// Whether the file is hidden from view in the OS
```
```suggestion
	/// Whether the file is hidden from view in the OS
	var/hidden = FALSE
```

SierraKomodo commented on code/modules/modular_computers/file_system/data.dm:
```diff
@@ -1,7 +1,7 @@
 // They don't contain other logic for now.
 /datum/computer_file/data
-	var/stored_data = "" 			// Stored data in string format.
+	var/stored_data = "" 			// Stored data in string format. Do not modify directly. Changes to file should be made through OS functions.
```
```suggestion
	/// Stored data in string format. Do not modify directly. Changes to file should be made through OS functions.
	var/stored_data = ""
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/antagonist/access_decrypter.dm:
```diff
@@ -16,7 +16,10 @@
 	var/datum/access/target_access = null
 	var/list/restricted_access_codes = list(access_change_ids) // access codes that are not hackable due to balance reasons
-	var/list/skill_restricted_access_codes_master = list(access_network)
+	var/list/skill_restricted_access_codes = list( // access codes that are restricted by skill
```
In this case, the var's name already tells us what the comment's saying.

```suggestion
	var/list/skill_restricted_access_codes = list(
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/antagonist/dos.dm:
```diff
@@ -105,16 +105,15 @@
-			ntnet_global.intrusion_detection_alarm = 1
-		return 1
+		ntnet_global.add_log_with_ids_check("Excess traffic flood targeting Quantum Relay ([target.uid]) detected from [length(sources_to_show)] device\s: [english_list(sources_to_show)]")
+		return 1
```
```suggestion
		return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/file_browser.dm:
```diff
@@ -27,37 +27,42 @@
-			F.filename = newname
+		if(!computer.rename_file(href_list["PRG_rename"], newname))
+			error = "File error: Unable to rename file."
+			return 1
```
```suggestion
			return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/file_browser.dm:
```diff
@@ -27,37 +27,42 @@
+			error = "File error: Unable to rename file."
+			return 1
+	if(href_list["PRG_usbdeletefile"])
+		. = 1
```
```suggestion
		. = TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/file_browser.dm:
```diff
@@ -27,37 +27,42 @@
+		. = 1
+		computer.delete_file(href_list["PRG_usbdeletefile"], computer.get_component(PART_DRIVE))
+	if(href_list["PRG_copytousb"])
+		. = 1
```
```suggestion
		. = TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/file_browser.dm:
```diff
@@ -27,37 +27,42 @@
+		. = 1
+		computer.copy_between_disks(href_list["PRG_copytousb"], computer.get_component(PART_HDD), computer.get_component(PART_DRIVE))
+	if(href_list["PRG_copyfromusb"])
+		. = 1
```
```suggestion
		. = TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/file_browser.dm:
```diff
@@ -27,37 +27,42 @@
+		. = 1
+		computer.copy_between_disks(href_list["PRG_copyfromusb"], computer.get_component(PART_DRIVE), computer.get_component(PART_HDD))
+	if(href_list["PRG_closefile"])
+		. = 1
```
```suggestion
		. = TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/file_browser.dm:
```diff
@@ -70,26 +75,19 @@
 		var/newtext = sanitize(replacetext(input(usr, "Editing file [open_file]. You may use most tags used in paper formatting:", "Text Editor", oldtext) as message|null, "\n", "\[br\]"), MAX_TEXTFILE_LENGTH)
 		if(!newtext)
-			return
+			return 1
```
```suggestion
			return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/reports.dm:
```diff
@@ -155,14 +155,14 @@
-		if(program.computer.store_file(file))
-			to_chat(user, "The report has been exported as [file.filename].[file.filetype]")
-		else
+		file.stored_data = selected_report.generate_pencode(get_access(user), no_html = 1) //TXT files can't have html; they use pencode only.
```
```suggestion
		file.stored_data = selected_report.generate_pencode(get_access(user), no_html = TRUE) //TXT files can't have html; they use pencode only.
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -18,15 +18,18 @@
 		open_file = F.filename
 		loaded_data = F.stored_data
-		return 1
+		is_edited = 0
```
```suggestion
		is_edited = FALSE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -58,38 +61,43 @@
-				save_file(open_file)
+				if(!save_file(open_file))
+					error = "I/O error: Unable to save file '[open_file]'."
+					browsing = 0
```
```suggestion
					browsing = FALSE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -58,38 +61,43 @@
+				if(!save_file(open_file))
+					error = "I/O error: Unable to save file '[open_file]'."
+					browsing = 0
+					return 1
```
```suggestion
					return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -58,38 +61,43 @@
 		browsing = 0
 		if(!open_file(href_list["PRG_openfile"]))
 			error = "I/O error: Unable to open file '[href_list["PRG_openfile"]]'."
+			return 1
```
```suggestion
			return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -58,38 +61,43 @@
-
+				if(!save_file(open_file))
+					error = "I/O error: Unable to save file '[open_file]'."
+					return 1
```
```suggestion
					return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -58,38 +61,43 @@
-			error = "I/O error: Unable to create file '[href_list["PRG_saveasfile"]]'."
+		open_file = F.filename
+		loaded_data = ""
+		return 1
```
```suggestion
		return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/wordprocessor.dm:
```diff
@@ -58,38 +61,43 @@
-			error = "I/O error: Unable to create file '[href_list["PRG_saveasfile"]]'."
+		if(!istype(F))
+			error = "I/O error: Unable to create file '[newname]'."
+			return 1
```
```suggestion
			return TRUE
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/research/email_administration.dm:
```diff
@@ -81,9 +88,13 @@
 
+	var/datum/extension/interactive/ntos/os = get_ntos()
+	if(!os)
+		return 1
```
```suggestion
		return TRUE
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -8,6 +8,7 @@
 	var/max_capacity = 128
 	var/used_capacity = 0
 	var/list/stored_files = list()		// List of stored files on this drive. DO NOT MODIFY DIRECTLY!
+	var/read_only = 0					// Whether drive is protected against changes
```
```suggestion
	/// Whether drive is protected against changes
	var/read_only = FALSE
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
-	store_file(new/datum/computer_file/program/computerconfig(src)) 		// Computer configuration utility, allows hardware control and displays more info than status bar
-	store_file(new/datum/computer_file/program/ntnetdownload(src))			// NTNet Downloader Utility, allows users to download more software from NTNet repository
-	store_file(new/datum/computer_file/program/filemanager(src))			// File manager, allows text editor functions and basic file manipulation.
+// Tries to find the file by filename. Returns null on failure
```
```suggestion
/// Tries to find the file by filename. Returns null on failure
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
-	store_file(new/datum/computer_file/program/ntnetdownload(src))			// NTNet Downloader Utility, allows users to download more software from NTNet repository
-	store_file(new/datum/computer_file/program/filemanager(src))			// File manager, allows text editor functions and basic file manipulation.
+// Tries to find the file by filename. Returns null on failure
+/obj/item/stock_parts/computer/hard_drive/proc/find_file_by_name(var/filename)
```
```suggestion
/obj/item/stock_parts/computer/hard_drive/proc/find_file_by_name(filename)
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
+		if(F.filename == filename)
+			return F
+
+// Use this proc to add Data file to the drive.
```
```suggestion
/// Use this proc to add Data file to the drive.
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
+			return F
+
+// Use this proc to add Data file to the drive.
+/obj/item/stock_parts/computer/hard_drive/proc/create_data_file(var/filename, var/data, var/file_type = /datum/computer_file/data, var/list/metadata)
```
```suggestion
/obj/item/stock_parts/computer/hard_drive/proc/create_data_file(filename, data, file_type = /datum/computer_file/data, list/metadata)
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
-/obj/item/stock_parts/computer/hard_drive/proc/remove_file(var/datum/computer_file/F)
-	if(!F || !istype(F))
-		return 0
+// Use this proc to save Data file to the drive.
```
```suggestion
/// Use this proc to save Data file to the drive.
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
-	if(!F || !istype(F))
-		return 0
+// Use this proc to save Data file to the drive.
+/obj/item/stock_parts/computer/hard_drive/proc/save_data_file(var/filename, var/data, var/file_type = /datum/computer_file/data, var/list/metadata)
```
```suggestion
/obj/item/stock_parts/computer/hard_drive/proc/save_data_file(filename, data, file_type = /datum/computer_file/data, list/metadata)
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
+	EF.calculate_size()
+	return save_file(EF)
+
+// Use this proc to add file to the drive. Contains necessary sanity checks.
```
```suggestion
/// Use this proc to add file to the drive. Contains necessary sanity checks.
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
+	return save_file(EF)
+
+// Use this proc to add file to the drive. Contains necessary sanity checks.
+/obj/item/stock_parts/computer/hard_drive/proc/create_file(var/datum/computer_file/F)
```
```suggestion
/obj/item/stock_parts/computer/hard_drive/proc/create_file(datum/computer_file/F)
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
+	recalculate_size()
+	return F
+
+// Use this proc to add file or update it if it already exists. Contains necessary sanity checks.
```
```suggestion
/// Use this proc to add file or update it if it already exists. Contains necessary sanity checks.
```

SierraKomodo commented on code/modules/modular_computers/hardware/hard_drive.dm:
```diff
@@ -60,107 +61,176 @@
+	return F
+
+// Use this proc to add file or update it if it already exists. Contains necessary sanity checks.
+/obj/item/stock_parts/computer/hard_drive/proc/save_file(var/datum/computer_file/F)
```
```suggestion
/obj/item/stock_parts/computer/hard_drive/proc/save_file(datum/computer_file/F)
```

SierraKomodo reviewed (changes requested):
Mostly formatting stuff.

I didn't go through every file, but you get the point - dmdocs, use TRUE/FALSE for booleans, and `var/` isn't needed in parameters.

Ithalan commented:
Going to make a pass over the stuff I've touched after getting some. Will try and fix up all the formatting issues that may remain there.

Ithalan commented:
Went through and cleaned up everything I could find that needed formatting in the files I've touched.

Couldn't resist the temptation to also improve a bit further. Connectivity for modular computers is now cached for the rest the current tick. This should protect against performance degradation even if some mad lad decides to set up every computer on the ship to run behind a hundred proxies.

Also fixed a bug that caused the revelations antag program to not work, and clicking the autorun icon for program that is already set to autorun will now remove the autorun flag.

SierraKomodo commented on code/modules/modular_computers/NTNet/NTNet.dm:
```diff
@@ -209,14 +215,13 @@ var/global/datum/ntnet/ntnet_global = new()
 			add_log("Configuration Updated. Wireless network firewall now [setting_systemcontrol ? "allows" : "disallows"] remote control of [station_name()]'s systems.")
 
-/datum/ntnet/proc/find_email_by_name(var/login)
+// Returns email account matching login. Otherwise null
```
```suggestion
/// Returns email account matching login. Otherwise null
```

SierraKomodo commented on code/modules/modular_computers/file_system/computer_file.dm:
```diff
@@ -1,17 +1,30 @@
-	var/undeletable = 0										// Whether the file may be deleted. Setting to 1 prevents deletion/renaming/etc.
-	var/uid													// UID of this file
-	var/list/metadata											// Any metadata the file uses.
+	// Placeholder. Whitespace and most special characters are not allowed.
```
```suggestion
	/// Placeholder. Whitespace and most special characters are not allowed.
```

SierraKomodo commented on code/modules/modular_computers/file_system/data.dm:
```diff
@@ -1,11 +1,13 @@
+
+	/// Stored data in string format. Do not modify directly. Changes to file should be made through OS functions.
+	var/stored_data = ""
+	/// Turns out, text is small.
```
This isn't really a comment that describes the var or its purpose.

Ithalan replied:
Fixed in coming commit

SierraKomodo commented on code/modules/modular_computers/file_system/program.dm:
```diff
@@ -1,31 +1,55 @@
+	var/datum/nano_module/NM = null
+	/// Path to nanomodule, make sure to set this if implementing new program.
+	var/nanomodule_path = null
+	/// PROGRAM_STATE_KILLED or PROGRAM_STATE_BACKGROUND or PROGRAM_STATE_ACTIVE - specifies whether this program is running.
```
Not required, but a little shorthand that I use in my comments to reduce the length of comments. VSC has autocomplete for defines and constants, so it's relatively simple to get a list of all `PROGRAM_STATE_*` matches.

```suggestion
	/// One of `PROGRAM_STATE_*` - specifies whether this program is running.
```

SierraKomodo commented on code/modules/modular_computers/file_system/program.dm:
```diff
@@ -1,31 +1,55 @@
-	var/ui_header = null							// Example: "something.gif" - a header image that will be rendered in computer's UI when this program is running at background. Images are taken from /nano/images/status_icons. Be careful not to use too large images!
-	var/ntnet_speed = 0								// GQ/s - current network connectivity transfer rate
-	var/operator_skill = SKILL_MIN                  // Holder for skill value of current/recent operator for programs that tick.
+	/// Bitflags (PROGRAM_CONSOLE, PROGRAM_LAPTOP, PROGRAM_TABLET, PROGRAM_PDA combination) or PROGRAM_ALL
```
As above.

```suggestion
	/// Bitflags (Any of `PROGRAM_*`) or PROGRAM_ALL
```

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/email_client.dm:
```diff
@@ -503,53 +502,58 @@
 
+		os = get_ntos()
+		if(!os)
+			return 1
```
```suggestion
			return TRUE
```

Ithalan replied:
Wasn't familiar enough with the Topic proc to tell if the return-value was only ever interpreted as a boolean, but I'll go through and change those too then.

SierraKomodo replied:
In hindsight I think there make actually be `TOPIC_*` defines used for it now but I'll have to dig and double check.

Ithalan replied:
All the files I touch should now be using TOPIC_* defines as return values where appropriate

SierraKomodo commented on code/modules/modular_computers/file_system/programs/generic/nttransfer.dm:
```diff
@@ -39,14 +39,20 @@ var/global/nttransfer_uid = 0
+/datum/computer_file/program/nttransfer/proc/get_ntnet_speed()
+	if(computer)
+		return computer.get_ntnet_speed(computer.get_ntnet_status())
+	return 0
```
```suggestion
	return FALSE
```

Ithalan replied:
get_ntnet_speed() is expected to return a decimal value presenting the transfer speed, so I don't think it's wrong to return zero even in an unlikely failure state.

SierraKomodo replied:
Right - I missed that was returning a numeric and not a boolean.

SierraKomodo commented on code/modules/modular_computers/file_system/reports/report.dm:
```diff
@@ -19,20 +29,20 @@
+  * Arguments can be access values (numbers) or lists of access values.
+  * If null is passed to one of the arguments, that access type is left alone. Pass list() to reset to no access needed instead.
+  * The recursive option resets access to all fields in the report as well.
+  * If the override option is set to 0, the access supplied will instead be added as another access pattern, rather than resetting the access.
```
Encourages usage of the defines instead of flat numbers for anyone reading the documentation.

```suggestion
  * If the override option is set to FALSE, the access supplied will instead be added as another access pattern, rather than resetting the access.
```

SierraKomodo commented on code/modules/modular_computers/hardware/_hardware.dm:
```diff
@@ -3,48 +3,59 @@
+
+	/// If the hardware uses extra power, change this.
+	var/power_usage = 0
+	/// If the hardware is turned off set this to 0.
```
```suggestion
	/// If the hardware is turned off set this to FALSE.
```

SierraKomodo commented on code/modules/modular_computers/hardware/network_card.dm:
```diff
@@ -49,75 +40,122 @@ var/global/ntnet_card_uid = 1
+	if(ethernet)
+		. += "OpenEth (Physical Connection) - Physical network connection port"
+
+/obj/item/stock_parts/computer/network_card/New(var/l)
+	..(l)
+	identification_id = ntnet_card_uid
+	ntnet_card_uid++
```
Use `Initialize()` instead of `New()` unless you absolutely have to use it.

```suggestion
/obj/item/stock_parts/computer/network_card/Initialize()
	. = ..()
	identification_id = ntnet_card_uid
	ntnet_card_uid++
```

SierraKomodo commented on code/modules/modular_computers/hardware/network_card.dm:
```diff
@@ -49,75 +40,122 @@ var/global/ntnet_card_uid = 1
+			return FALSE
+	return TRUE
+
+/// Returns a list of all network cards this connection passes through before terminating
+/obj/item/stock_parts/computer/network_card/proc/get_route(list/routed_through) // Argument is a safety parameter for internal calls. Don't use manually.
```
```suggestion
/// Returns a list of all network cards this connection passes through before terminating. Argument is a safety parameter for internal calls. Don't use manually.
/obj/item/stock_parts/computer/network_card/proc/get_route(list/routed_through)
```

SierraKomodo commented on code/modules/modular_computers/hardware/portable_hard_drive.dm:
```diff
@@ -31,11 +31,12 @@
 /obj/item/stock_parts/computer/hard_drive/portable/merchant
 	name = "merchant_list_1155_CRACKZ_1155_no_keygen_repack"
 	desc = "An obviously pirated copy of well-known trading software."
+	read_only = 1
```
```suggestion
	read_only = TRUE
```

SierraKomodo commented on code/modules/modular_computers/hardware/portable_hard_drive.dm:
```diff
@@ -1,4 +1,4 @@
-// These are basically USB data sticks and may be used to transfer files between devices
+/// These are basically USB data sticks and may be used to transfer files between devices
```
As a note, dmdocs don't parse and aren't needed for type paths - Only for initial `/proc` and `/var` definitions - And for `/var`, only for properties of a given type or globals.

Ithalan replied:
Will keep in mind for the future. Leaving these as is for now however.

SierraKomodo commented on code/modules/modular_computers/hardware/scanners/scanner.dm:
```diff
@@ -1,19 +1,23 @@
+	var/can_run_scan = FALSE
+	/// Whether the scan output can be viewed in the program.
+	var/can_view_scan = TRUE
+	//Whether the scan output can be saved to disk.
```
```suggestion
	/// Whether the scan output can be saved to disk.
```

SierraKomodo commented on code/modules/modular_computers/ntos/ntos.dm:
```diff
@@ -2,20 +2,30 @@
+	var/menu_icon = "menu"
 	var/screensaver_icon = "standby"
 
+	// Cached NTNet status and network tag for outgoing connections to avoid checking entire proxy chain every time
+	var/ntnet_status
+	var/network_tag
```
```suggestion
	/// Cached NTNet status for outgoing connections to avoid checking entire proxy chain every time
	var/ntnet_status
	/// Cached NTNet network tag for outgoing connections to avoid checking entire proxy chain every time
	var/network_tag
```

SierraKomodo commented on code/modules/modular_computers/ntos/ntos.dm:
```diff
@@ -2,20 +2,30 @@
 	var/list/last_header_icons
 
-	//Pain and suffering
+	// Pain and suffering. Sometimes computers are made to update, and you just have to sit through or risk damage by rebooting it
```
```suggestion
	/// Pain and suffering. Sometimes computers are made to update, and you just have to sit through or risk damage by rebooting it
```

SierraKomodo reviewed (commented):
Regarding access:
- I'm not sure why Roboticist needs access to network stuff.
- I can see why Forensics Technician could use the access, but as someone who plays FT I've found that lacking the access encourages cooperating with command/engineering to get the data needed for an investigation (Specifically using `locate` and I can see the new `log` command also being useful). Inter-departmental stuff is good.

Ithalan commented:
Roboticist already start with a minimum of basic in Information Technology, so it seemed sensible for them to be among those trusted with basic network access.

The general idea of the basic network access was to give roles that invest in the Information Technology skill and also have an  even vaguely valid reason for accessing or monitoring the network the ability to do so, with automatic oversight where appropriate to allow misuse to be held accountable. All of the access-locked commands, with the exception of Banned, already requires Experienced or better in the skill, so there's no guarantee that any one person with the access will also be able to put the access to good use.

Ithalan committed:
Adds new terminal commands and overhauls parts of NTOS

SierraKomodo reviewed (changes requested):
Tested each of the commands, two things:

- The copy function of the `file` command deletes the original file when making the copy.
- UX request - `hwinfo` output should include the name of the device it's returning information on. This is rather ambiguous and non descript:
```
> hwinfo nt
hwinfo: Running diagnostic protocols...
Hardware Integrity Test... (Corruption: 0/100) PASS
> hwinfo ntnet
hwinfo: Running diagnostic protocols...
Hardware Integrity Test... (Corruption: 0/100) PASS
NIX Unique ID: 66
NIX User Tag: 
Supported protocols:
511.m SFS (Subspace) - Standard Frequency Spread
OpenEth (Physical Connection) - Physical network connection port
```

Ithalan committed:
Adds device name to hwinfo and fixes missing copy flag in file command

Ithalan commented:
Both changed as requested.
